File: bot-api/dotnet/src/mapper/BotStateMapper.cs
namespace Robocode.TankRoyale.BotApi.Mapper
{
  public sealed class BotStateMapper
  {
    public static BotState Map(Schema.BotState source)
    {
      return new BotState(
        source.Energy,
        source.X,
        source.Y,
        source.Direction,
        source.GunDirection,
        source.RadarDirection,
        source.RadarSweep,
        source.Speed,
        source.TurnRate,
        source.GunTurnRate,
        source.RadarTurnRate,
        source.GunHeat,
        Color.FromRgbInt(source.BodyColor),
        Color.FromRgbInt(source.TurretColor),
        Color.FromRgbInt(source.RadarColor),
        Color.FromRgbInt(source.BulletColor),
        Color.FromRgbInt(source.ScanColor),
        Color.FromRgbInt(source.TracksColor),
        Color.FromRgbInt(source.GunColor)
      );
    }
  }
}

File: bot-api/typescript/src/mapper/BotStateMapper.ts
import { BotState as SchemaBotState } from "../protocol/schema.js";
import { BotState } from "../BotState.js";

/** Maps schema BotState to API BotState. */
export class BotStateMapper {
  static map(s: SchemaBotState): BotState {
    return new BotState(
      s.isDroid,
      s.energy,
      s.x,
      s.y,
      s.direction,
      s.gunDirection,
      s.radarDirection,
      s.radarSweep,
      s.speed,
      s.turnRate,
      s.gunTurnRate,
      s.radarTurnRate,
      s.gunHeat,
      s.enemyCount,
      s.bodyColor ?? null,
      s.turretColor ?? null,
      s.radarColor ?? null,
      s.bulletColor ?? null,
      s.scanColor ?? null,
      s.tracksColor ?? null,
      s.gunColor ?? null,
      s.isDebuggingEnabled,
    );
  }
}
